### DISCUSSION TITLE
DAI 2023-2024 - Check out my application protocol! - Lucas Lattion

### DISCUSSION THREAD

**Discussion by lucaslattion** [heig-vd-dai-course/heig-vd-dai-course#106] [Show and tell]
**Purpose of the Protocol:**
The purpose of this protocol is to enable users to send text messages to each other via a central server. It facilitates the exchange of messages in a client-server architecture.

**Port and Protocol:**
- Port: 1234
- Protocol: TCP (Transmission Control Protocol)

**Connection Initiation:**
- Clients initiate the connection to the server running on port 1234.

**Available Messages/Actions:**
1. **Connect:** A client initiates a connection request to the server to establish communication. The server can accept or refuse the connection based on its capacity.

2. **Send Message:** Once connected, users can send text messages to other connected users.

**Message/Action Formats:**
1. **Connect Request:**
   - Format: `CONNECT <username>`
   - Example: `CONNECT Alice`
   - Description: A client sends a `CONNECT` request to the server with a unique username.

2. **Connect Response:**
   - Format: `CONNECTED <username>`
   - Example: `CONNECTED Alice`
   - Description: The server responds with a `CONNECTED` message, indicating a successful connection with the provided username.

3. **Refuse Connection:**
   - Format: `REFUSED`
   - Example: `REFUSED`
   - Description: If the server is at maximum capacity (5 connections), it responds with a `REFUSED` message, indicating that the connection request is denied.

4. **Send Message:**
   - Format: `SEND <recipient> <message>`
   - Example: `SEND Bob Hello, Bob!`
   - Description: A connected user sends a `SEND` request to the server, specifying the recipient's username and the message to be sent.

5. **Message Delivered:**
   - Format: `DELIVERED <sender> <message>`
   - Example: `DELIVERED Alice Hello, Bob!`
   - Description: The server delivers a message to the intended recipient. Both the sender and the message content are included in the response.

**Edge Cases and Error Handling:**

- **Connection Refusal:** If the server is at maximum capacity (5 connections), it refuses any new connection requests with a `REFUSED` message.
- **Message Length Limitation:** Messages sent between users must not exceed a certain length (e.g., 100 characters). If a message exceeds this limit, an error response should be sent back to the sender.
- **Invalid Actions:** If a client sends an invalid action or message format, the server should respond with an error message indicating the action or message format is not recognized.

Here's a simplified PlantUML sequence diagram to represent the protocol:

```plantuml
@startuml
Client -> Server: CONNECT Alice
Server --> Client: CONNECTED Alice
Client -> Server: SEND Bob Hello, Bob!
Server --> Client: DELIVERED Alice Hello, Bob!
@enduml
```
![DiagramProtocolDAI-1](https://github.com/heig-vd-dai-course/heig-vd-dai-course/assets/71433770/b9deb7a5-ec1f-4a89-a99a-f76c805a427d)

This sequence diagram illustrates the typical flow of connecting and sending a message between a client (Alice) and the server, with appropriate responses.

Tools used: ChatGPT, PlantUML

**Comment by ludelafo** [comment]
Great!

Adding examples is very nice to help understand your protocol.

I would also use enums/constants for the edge cases and error handling to describe the protocol, such as it is done in the message/action formats.

I would also add more examples in the sequence diagram to illustrate all use cases. The more details your application protocol contains, the less error-prone it will be.

P.S.: In the current description of your application protocol, two users must be connected at the same time to receive/send text messages, is that right? You could add another command to retrieve messages that were sent while being offline, such as `RETRIEVE`. Do you see what I mean?
